rpi_pico: Add PIO based WS2812 LED strip driver

soburi opened the pull request:
This PR adds a WS2812 driver that uses the PIO functionality of the Raspberry Pi Pico.

#44316 already merged. I rebased this PR for newest.

<strike>
Support for PIO functionality is already in progress in #44316, and this PR captures some of that commits.

This demonstrates a different PIO use case than #44316.
I hope it will help discussion about implementing PIO function.

---

- Suggestions for PIO driver layer

This PR  proposes an extension to make the GPIO pins more configurable by adding the PIO driver that configures which pin the PIO uses.
The PIO driver mainly manages resources, and specific functions are implemented by each client driver.
Specifically, add two properties, `gpios` and `gpio-names` to the PIO node of dts so that you can specify the pin by name. These properties allow client drivers to look up pins by symbolic name, allow to change pin assignments without driver modification.

```
&pio0 {
        status = "okay";
        pinctrl-0 = <&ws2812_pio0_default>;
        pinctrl-names = "default";

        /* declare how many state machine uses */
        state-machine-num = <1>;

        /* declare gpio pins and that names */
        gpios = <&gpio0 17 0>;
        gpio-names = "ws2812";

        led_strip: ws2812 {
                compatible = "worldsemi,ws2812-rpi_pico-pio";
                status = "okay";
                chain-length = <1>;
                color-mapping = <LED_COLOR_ID_GREEN
                                 LED_COLOR_ID_RED
                                 LED_COLOR_ID_BLUE>;
        };
};
```


---

@yonsch 
I made it based on your PR.
I would be happy if you could give me your opinion.

</strike>

soburi committed:
drivers: led_strip: add rpi_pico's PIO based ws2812 driver

Add driver that based on RPI-PICO's PIO feature for ws2812.

This driver can handle WS2812 or compatible LED strips.
The single PIO node can handle up to 4 strips.
Any pins that can be configured for PIO can be used for strips.

I verified the samples/driver/led_ws2812 sample
working with WS2812(144 pcs) led strip using following patches.

- samples/drivers/led_ws2812/boards/rpi_pico.overlay

```
/ {
        aliases {
                led-strip = &ws2812;
        };
};

&pinctrl {
        ws2812_pio0_default: ws2812_pio0_default {
                ws2812 {
                        pinmux = <PIO0_P21>;
                };
        };
};

&pio0 {
        status = "okay";

        pio-ws2812 {
                compatible = "worldsemi,ws2812-rpi_pico-pio";
                status = "okay";
                pinctrl-0 = <&ws2812_pio0_default>;
                pinctrl-names = "default";
                bit-waveform = <3>, <3>, <4>;

                ws2812: ws2812 {
                        status = "okay";
                        output-pin = <21>;
                        chain-length = <144>;
                        color-mapping = <LED_COLOR_ID_GREEN
                                         LED_COLOR_ID_RED
                                         LED_COLOR_ID_BLUE>;
                        reset-delay = <280>;
                        frequency = <800000>;
                };
        };
};

```

- samples/drivers/led_ws2812/boards/rpi_pico.conf

```
CONFIG_WS2812_STRIP_RPI_PICO_PIO=y
```

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
boards: arm: adafruit_kb2040: add led-strip configuration

Adafruit KB2040 has one NeoPixel(WS2812) LED that
attaches to GPIO17 pin.
Add configuration for it.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
samples: drivers: led_ws2812: add configure for adafruit_kb2040

Add configuration to use PIO driver for adafruit_kb2040.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

yonsch commented:
Thanks @soburi , I really want to get PIO merged before v3.4.
I like your proposal of creating a "base" PIO driver and extending it in the protocol drivers.
There was also a proposal to include pioasm somehow and assemble the programs on every build, but I think this feature can be added later if we choose to accept it.
I hope to get your changes merged into mine this week, and we can start moving this forward.

yonsch commented on boards/arm/adafruit_kb2040/adafruit_kb2040.dts:
```diff
@@ -95,6 +98,25 @@
+	pinctrl-names = "default";
+	state-machine-num = <1>;
+
+	gpios = <&gpio0 17 0>;
```
My problem with this is that it requires you to redefine the pin here (as 17), when you've already done so in the pinctrl state. Now to change the pin you have to change both. Ideally, pinctrl would allow us to define names for pins in the pinctrl state, but currently it isn't supported. Any other ideas?

soburi replied:
IMO, there isn't a better way at this point.
Of course, I also considered using pinctrl, but there were some difficulties.

First, pinctrl has a hierarchical structure.
Creating a macro that statically resolves name-to-pin correspondences takes a lot of work.
(It should change dts parsing script.)

We also need to determine the valid scope definitions for names.
Also, pinctrl can switch between sets of settings depending on the state. (Only default can select for rpi_pico at this time.) 
Since the changed setting may not always include the definition of the pin name, it is necessary to consider the behavior at that time.

Given the above, this is out of the current pinctrl design scope, so I'm choosing the current method.
It would be a good idea to add pin names in pinctrl, but it's a too big issue to address in this PR.

yonsch commented on drivers/misc/pio_rpi_pico/pio_rpi_pico.c:
```diff
@@ -0,0 +1,97 @@
+	};                                                                                         \
+                                                                                                   \
+	DEVICE_DT_INST_DEFINE(idx, &pio_rpi_pico_init, NULL, &pio_rpi_pico_data_##idx,             \
+			      &pio_rpi_pico_config_##idx, PRE_KERNEL_2,                            \
```
Is there a reason for choosing PRE_KERNEL_2 and not earlier? If possible, I think it should be initialized as early as possible, because devices relying on this in the future might need it to be ready very early

soburi replied:
It makes sense.
I change it to PRE_KERNEL_1, the same as the reset-controller priority.
It is early enough before other driver initialization.

yonsch reviewed (commented):
I'm integrating your proposal into my PR, I added some questions.
Another general question: Can two drivers use the same PIO given that it has enough state machines. If it can't, can we add an assertion to make sure two drivers aren't trying to use the same PIO?

soburi commented:
@yonsch 

> I'm integrating your proposal into my PR, I added some questions.
> Another general question: Can two drivers use the same PIO given that it has enough state machines. If it can't, can we add an assertion to make sure two drivers aren't trying to use the same PIO?

Added an assert to error if there are more than 2 child nodes.
As a result, protocol drivers can now occupy PIOs.
Managing the PIO driver's state machine is redundant and should be removed.

simonguinot commented on dts/bindings/led_strip/worldsemi,ws2812-rpi_pico-pio.yaml:
```diff
@@ -0,0 +1,20 @@
+properties:
+  frequency:
+    type: int
+    default: 800000
```
Does the driver support a different frequency ? I don't see in the driver how the signal is adjusted according to the frequency.

simonguinot replied:
Yes but then changing the frequency also changes the duration of the pulses output by the PIO program. AFAIU (and I don't understand much :) the pulses seems to be "hardcoded" by the PIO program instructions. So I suspect that using a different frequency won't work.

mbolivar-nordic replied:
I see -- if that's the case, there seems to be no need for this property.

soburi replied:
This property intends  to supporting WS2811(400kHz), WS2813 or WS2815(2MHz).
It is inherit from https://github.com/raspberrypi/pico-examples/tree/master/pio/ws2812, not yet tested excluding 800000 in zephyr.

simonguinot replied:
So how does this frequency affect the pulses ?

Let's say a user selects the 400kHz frequency for a strip of WS2811 devices. In that case is it still correct for the PIO program to output `0001100000` and `0001111111` pulses for 0 and 1 bits ?

Did you consider to make the pulses configurable as in the SPI driver (see the `spi-one-frame` and `spi-zero-frame `DT properties) ? This would make this driver compatible with more LED controllers.

soburi replied:
I understood it that is the optimal value is different by LED models.
But I'm not have other than ws2812.
I'm ordering ws2813 now. I'll consider to add the option for tune to particular LED model.

simonguinot replied:
So how does this frequency affect the pulses ?

Let's say a user selects the 400kHz frequency for a strip of WS2811 devices. In that case is it still correct for the PIO program to output `0001100000` and `0001111111` pulses for 0 and 1 bits ?

Did you consider to make the pulses configurable as in the SPI driver (see the `spi-one-frame` and `spi-zero-frame` DT properties) ? This would make this driver compatible with more LED controllers.

simonguinot replied:
> I understood it that is the optimal value is different by LED models. But I'm not have other than ws2812. I'm ordering ws2813 now. I'll consider to add the option for tune to particular LED model.

The frequency "probably" affects the width pulse of a bit. So I don't understand how hardcoded pulse values for 0 and 1 bits can work with different frequencies. Indeed it would be interesting to test with another model.

simonguinot reviewed (changes requested):
Hi @soburi,

The Raspberry Pi PIO interface looks really great. And it is nice to see how it can be used to handle WS2812 and compatible LED strip controllers.

In addition your driver looks good. Please see my comments and questions below.

simonguinot commented on boards/arm/adafruit_kb2040/adafruit_kb2040-pinctrl.dtsi:
```diff
@@ -47,4 +47,10 @@
 			input-enable;
 		};
 	};
+
```
And please add a description as well to commit adding the driver. It would be useful to describe a bit the PIO interface and how it can be used to output a signal compatible with WS2812 LED strip controllers.

soburi replied:
I added information to commit message about setting that used to verify this patch.

zpm1066 commented:
Hi @soburi,

I have an `Adafruit RP2040` with a builtin RGB NeoPixel. 

Currently, I'm running a 8 NeoPixel led strip with the Adafruit RP2040 using SPI. I'd to drive the builtin NeoPixel but it appears that the current led_strip driver only works with nRF devices. 

Would I be able use your PIO based `WS2812 LED driver` to control the [Adafruit RP2040](url) builtin NeoPixel?

I tried using the following but ran into errors.  

Any suggestions on how I can resolve the issue(s)? Thanks.


`pinctrl in the RP2040 rip_pico.overlay`

```
&pinctrl {
	spi0_default: spi0_default {
		group1 {
			pinmux = <SPI0_CSN_P17>, <SPI0_SCK_P18>, <SPI0_TX_P19>;
		};
		group2 {
			pinmux = <SPI0_RX_P16>;
			input-enable;
		};
	};
	
	ws2812_pio0_default: ws2812_pio0_default {
		ws2812 {
			pinmux = <PIO0_P16>;
		};
	};
};

```

`WS2812 section in prj.conf`

```
# WS2812
CONFIG_SPI=n
CONFIG_WS2812_STRIP_SPI=n
CONFIG_WS2812_STRIP_RPI_PICO_PIO=y
CONFIG_WS2812_STRIP=y
CONFIG_LED_STRIP=y
CONFIG_LED_STRIP_LOG_LEVEL_DBG=y
```

`Build errors:`

```
(zephyrenv) quark11:feather_rp2040_neopixel_ws2812_pio quark11$ west build -b rpi_pico -p auto
[0/1] Re-running CMake...
Loading Zephyr default modules (Zephyr base (cached)).
-- Application: /Users/quark11/Documents/2022_zephyr_dev/2022_zephyr_pico_dev/feather_rp2040_neopixel_ws2812_pio
-- CMake version: 3.26.3
-- Cache files will be written to: /Users/quark11/Library/Caches/zephyr
-- Zephyr version: 3.3.99 (/Users/quark11/Downloads/zephyr-pico_ws2812/zephyr)
-- Found west (found suitable version "0.12.0", minimum required is "0.7.1")
-- Board: rpi_pico
-- Found toolchain: gnuarmemb (/usr/local/)
-- Found BOARD.dts: /Users/quark11/Downloads/zephyr-pico_ws2812/zephyr/boards/arm/rpi_pico/rpi_pico.dts
-- Found devicetree overlay: /Users/quark11/Documents/2022_zephyr_dev/2022_zephyr_pico_dev/feather_rp2040_neopixel_ws2812_pio/boards/rpi_pico.overlay
-- Generated zephyr.dts: /Users/quark11/Documents/2022_zephyr_dev/2022_zephyr_pico_dev/feather_rp2040_neopixel_ws2812_pio/build/zephyr/zephyr.dts
-- Generated devicetree_generated.h: /Users/quark11/Documents/2022_zephyr_dev/2022_zephyr_pico_dev/feather_rp2040_neopixel_ws2812_pio/build/zephyr/include/generated/devicetree_generated.h
-- Including generated dts.cmake file: /Users/quark11/Documents/2022_zephyr_dev/2022_zephyr_pico_dev/feather_rp2040_neopixel_ws2812_pio/build/zephyr/dts.cmake
/Users/quark11/Documents/2022_zephyr_dev/2022_zephyr_pico_dev/feather_rp2040_neopixel_ws2812_pio/build/zephyr/zephyr.dts:71.34-76.5: Warning (simple_bus_reg): /soc/peripheral-clk: missing or empty reg/ranges property
/Users/quark11/Documents/2022_zephyr_dev/2022_zephyr_pico_dev/feather_rp2040_neopixel_ws2812_pio/build/zephyr/zephyr.dts:77.26-82.5: Warning (simple_bus_reg): /soc/system-clk: missing or empty reg/ranges property
Parsing /Users/quark11/Downloads/zephyr-pico_ws2812/zephyr/Kconfig
Loaded configuration '/Users/quark11/Documents/2022_zephyr_dev/2022_zephyr_pico_dev/feather_rp2040_neopixel_ws2812_pio/build/zephyr/.config'
No change to configuration in '/Users/quark11/Documents/2022_zephyr_dev/2022_zephyr_pico_dev/feather_rp2040_neopixel_ws2812_pio/build/zephyr/.config'
No change to Kconfig header in '/Users/quark11/Documents/2022_zephyr_dev/2022_zephyr_pico_dev/feather_rp2040_neopixel_ws2812_pio/build/zephyr/include/generated/autoconf.h'

CMake Warning at /Users/quark11/Downloads/zephyr-pico_ws2812/zephyr/CMakeLists.txt:838 (message):
  No SOURCES given to Zephyr library: drivers__led_strip

  Excluding target from build.


-- Configuring done (4.6s)
-- Generating done (0.2s)
-- Build files have been written to: /Users/quark11/Documents/2022_zephyr_dev/2022_zephyr_pico_dev/feather_rp2040_neopixel_ws2812_pio/build
[1/140] Performing build step for 'second_stage_bootloader'
ninja: no work to do.
[128/138] Linking C executable zephyr/zephyr_pre0.elf
FAILED: zephyr/zephyr_pre0.elf zephyr/zephyr_pre0.map /Users/quark11/Documents/2022_zephyr_dev/2022_zephyr_pico_dev/feather_rp2040_neopixel_ws2812_pio/build/zephyr/zephyr_pre0.map 
: && ccache /usr/local/bin/arm-none-eabi-gcc  -gdwarf-4 zephyr/CMakeFiles/zephyr_pre0.dir/misc/empty_file.c.obj -o zephyr/zephyr_pre0.elf  zephyr/CMakeFiles/offsets.dir/./arch/arm/core/offsets/offsets.c.obj  -fuse-ld=bfd  -Wl,-T  zephyr/linker_zephyr_pre0.cmd  -Wl,-Map=/Users/quark11/Documents/2022_zephyr_dev/2022_zephyr_pico_dev/feather_rp2040_neopixel_ws2812_pio/build/zephyr/zephyr_pre0.map  -Wl,--whole-archive  app/libapp.a  zephyr/libzephyr.a  zephyr/arch/common/libarch__common.a  zephyr/arch/arch/arm/core/aarch32/libarch__arm__core__aarch32.a  zephyr/arch/arch/arm/core/aarch32/cortex_m/libarch__arm__core__aarch32__cortex_m.a  zephyr/lib/libc/minimal/liblib__libc__minimal.a  zephyr/lib/libc/common/liblib__libc__common.a  zephyr/lib/posix/liblib__posix.a  zephyr/soc/arm/rpi_pico/rp2/libsoc__arm__rpi_pico__rp2.a  zephyr/drivers/console/libdrivers__console.a  zephyr/drivers/gpio/libdrivers__gpio.a  zephyr/drivers/pinctrl/libdrivers__pinctrl.a  zephyr/drivers/reset/libdrivers__reset.a  zephyr/drivers/serial/libdrivers__serial.a  zephyr/drivers/timer/libdrivers__timer.a  modules/hal_rpi_pico/libmodules__hal_rpi_pico.a  modules/segger/libmodules__segger.a  -Wl,--no-whole-archive  zephyr/kernel/libkernel.a  -L"/usr/local/Cellar/arm-none-eabi-gcc/10.3-2021.07/gcc/bin/../lib/gcc/arm-none-eabi/10.3.1/thumb/v6-m/nofp"  -L/Users/quark11/Documents/2022_zephyr_dev/2022_zephyr_pico_dev/feather_rp2040_neopixel_ws2812_pio/build/zephyr  -lgcc  zephyr/arch/common/libisr_tables.a  -mcpu=cortex-m0plus  -mthumb  -mabi=aapcs  -Wl,--gc-sections  -Wl,--build-id=none  -Wl,--sort-common=descending  -Wl,--sort-section=alignment  -Wl,-u,_OffsetAbsSyms  -Wl,-u,_ConfigAbsSyms  -nostdlib  -static  -Wl,-X  -Wl,-N  -Wl,--orphan-handling=warn && cd /Users/quark11/Documents/2022_zephyr_dev/2022_zephyr_pico_dev/feather_rp2040_neopixel_ws2812_pio/build/zephyr && /usr/local/Cellar/cmake/3.26.3/bin/cmake -E echo
/usr/local/Cellar/arm-none-eabi-gcc/10.3-2021.07/gcc/bin/../lib/gcc/arm-none-eabi/10.3.1/../../../../arm-none-eabi/bin/ld.bfd: app/libapp.a(main.c.obj): in function `main':
/Users/quark11/Documents/2022_zephyr_dev/2022_zephyr_pico_dev/feather_rp2040_neopixel_ws2812_pio/src/main.c:63: undefined reference to `__device_dts_ord_54'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /usr/local/bin/cmake --build /Users/quark11/Documents/2022_zephyr_dev/2022_zephyr_pico_dev/feather_rp2040_neopixel_ws2812_pio/build
(zephyrenv) quark11:feather_rp2040_neopixel_ws2812_pio quark11$ 

```

simonguinot commented:
> I have an `Adafruit RP2040` with a builtin RGB NeoPixel.
> 
> Currently, I'm running a 8 NeoPixel led strip with the Adafruit RP2040 using SPI. I'd like to drive the builtin NeoPixel but it appears that the current led_strip driver only works with nRF devices.

It is not @zpm1066. I have been using the ws2812_spi driver with various ST and NXP devices.

zpm1066 commented:
> > I have an `Adafruit RP2040` with a builtin RGB NeoPixel.
> > Currently, I'm running a 8 NeoPixel led strip with the Adafruit RP2040 using SPI. I'd like to drive the builtin NeoPixel but it appears that the current led_strip driver only works with nRF devices.
> 
> It is not @zpm1066. I have been using the ws2812_spi driver with various ST and NXP devices.

Hi @simonguinot. Thanks. Yes, the ws2812_spi driver works well with nRF52, STM, and RP2040 devices.
Adafruit’s Arduino core ws2812 driver works both with built-in and SPI. I’ll dig into that to see how it works for their built-in neopxel.

mbolivar-nordic reviewed (changes requested):
Thanks for the driver!

I'm definitely interested in getting more ws2812 drivers merged into Zephyr. They are such weird little devices and I think having SoC-specific drivers when possible that use nice features like pio makes total sense.

A few comments and questions from my side. I also agree with everything @simonguinot has posted so far.

soburi commented on drivers/led_strip/ws2812_rpi_pico_pio.c:
```diff
@@ -0,0 +1,158 @@
+				color |= 0;
+				break;
+			case LED_COLOR_ID_RED:
+				color |= pixels[i].r << (8 * (2 - j));
+				break;
+			case LED_COLOR_ID_GREEN:
+				color |= pixels[i].g << (8 * (2 - j));
+				break;
+			case LED_COLOR_ID_BLUE:
+				color |= pixels[i].b << (8 * (2 - j));
+				break;
```
The `j` is from index in color_mapping array. It takes from 0, 1 or 2.
So, if color_mapping is [RED, GREEN, BLUE], 
the calculation is (pixels[i].r << (8 * (2)) | pixels[i].g << (8 * (1) | pixels[i].b << (8 * (0)).
Otherwise,  color_mapping is [GREEN, RED, BLUE] case,
the calculation is (pixels[i].r << (8 * (1)) | pixels[i].g << (8 * (2) | pixels[i].b << (8 * (0)).

It sort the bits correctly based on the color_mapping settings.

soburi commented on drivers/led_strip/Kconfig.ws2812:
```diff
@@ -46,4 +46,11 @@ config WS2812_STRIP_GPIO
 	  controller.
 
+config WS2812_STRIP_RPI_PICO_PIO
+	bool "Raspberry Pi PIO WS2812 led strip driver"
```
Thank you for reviewing.

I fixed it.

soburi commented on drivers/led_strip/ws2812_rpi_pico_pio.c:
```diff
@@ -0,0 +1,158 @@
+	const struct pinctrl_dev_config *pcfg;
+	const uint32_t pin;
+	const uint8_t num_colors;
+	const uint32_t frequency;
```
Fixed it.

soburi replied:
Fixed it.

soburi commented on boards/arm/adafruit_kb2040/adafruit_kb2040.dts:
```diff
@@ -95,6 +97,21 @@
+		chain-length = <1>;
+		color-mapping = <LED_COLOR_ID_GREEN
+				 LED_COLOR_ID_RED
+				 LED_COLOR_ID_BLUE>;
```
Fixed it.

soburi commented on drivers/led_strip/ws2812_rpi_pico_pio.c:
```diff
@@ -0,0 +1,158 @@
+			}
+		}
+
+		pio_sm_put_blocking(pio, 0, color << 8);
```
> I don't know how works PIO but in ws2812_rpi_pico_pio_init() you initialized a state machine returned by pio_claim_unused_sm(). So is it correct to use 0 here (which is eventually not the SM you initialized) ?

You are absolutely right. It my mistake.
Fixed it.

soburi replied:
`pio_sm_put_blocking` blocks only when fifo is full. otherwise, append data to last of fifo and returns.

simonguinot commented on dts/bindings/led_strip/worldsemi,ws2812-rpi_pico-pio.yaml:
```diff
@@ -0,0 +1,20 @@
+
+compatible: "worldsemi,ws2812-rpi_pico-pio"
+
+include: [base.yaml, ws2812.yaml, "raspberrypi,pico-pio-device.yaml"]
```
Hi @mbolivar-nordic. Having different bindings for a same compatible based on the bus type is a nice feature that I was not aware of. Yes I agree it would be a bit cleaner. For the I2C and GPIO unified binding how does it work to require a property OR another ?

soburi replied:
This is a widely effect topic.
I think it is good to discuss in separate topic.

simonguinot commented:
Hi @soburi. Please can you give us an update on this PR ?

soburi commented:
@simonguinot 
A little time is taken up by other work. Will update soon.

simonguinot commented:
Hi @soburi. Do you still intent to get this PR merged ?

soburi commented on drivers/led_strip/ws2812_rpi_pico_pio.c:
```diff
@@ -0,0 +1,158 @@
+	0x1400, /*  2: jmp    0               side 1 [4]  */
+	0xa442, /*  3: nop                    side 0 [4]  */
+		/*  .wrap                                 */
+);
```
Added comments.

yonsch reviewed (dismissed):
I'm unfamiliar with the LED but the PIO parts look fine

simonguinot commented:
Hi @soburi. Thanks for resuming this PR. I'll take a look at it this weekend.

soburi commented:
> Hi @soburi. Thanks for resuming this PR. I'll take a look at it this weekend.

Thank you in advance.
In actually, I just verified that there was no problem by looking at the specifications and implementation of the points I received before. The code is not changed.

simonguinot commented on drivers/led_strip/ws2812_rpi_pico_pio.c:
```diff
@@ -0,0 +1,194 @@
+			}
+		}
+
+		pio_sm_put_blocking(pio, data->sm, color << 8);
```
How is handled this third byte set to 0 ? Won't it be interpreted by the next controller in the strip as the value of its first color ?

simonguinot replied:
In fact, it won't. Let me clarify my thinking: do we emit extra zeros on the wire (one byte out of four) ? And if so, can we optimize the output data in order to avoid that ?

soburi replied:
https://github.com/zephyrproject-rtos/zephyr/pull/55226/files#diff-e3c7e6a4007d7de0939e6d3409d542a944cd2771cb6db37fb806945f3989a257R56

The process at this line enforcing uses only 24bit for actual transmission.
The last 1 byte will discard.

soburi replied:
Improvement:
Detect 24bit/32bit color from the `color-mapping` configuration, applying it to sending bits.

simonguinot commented on drivers/led_strip/ws2812_rpi_pico_pio.c:
```diff
@@ -0,0 +1,194 @@
+static int ws2812_rpi_pico_pio_init(const struct device *dev)
+{
+	const struct ws2812_rpi_pico_pio_config *config = dev->config;
+	struct ws2812_rpi_pico_pio_data *data = dev->data;
```
Shouldn't we check here if the PIO device is ready ?

```
 if (!device_is_ready(config->piodev)) {
        LOG_ERR("%s: PIO device not ready", dev->name);
        return -ENODEV;
}
```

soburi replied:
It's make a sense.
I applied it. Thank you!

simonguinot commented on dts/bindings/led_strip/worldsemi,ws2812-rpi_pico-pio.yaml:
```diff
@@ -0,0 +1,21 @@
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+  Worldsemi WS2812 LED strip, RaspberryPi Pico's PIO binding
```
I am renewing my question: is the PIO mechanism only available on the RaspberryPi Pico device ? If not then maybe we can remove the "pico" string from the driver name and from other places in DT ?

soburi replied:
> is the PIO mechanism only available on the RaspberryPi Pico device 

Yes it is. PIO is only available on Rpi-Pico.
But "PIO" is easy to conflict name. I think it is better to clarify the context.

simonguinot reviewed (commented):
Hi @soburi,

Thanks for this update. This driver is quite straightforward and it is looking good to me.

I only have few questions left. Please see them below.

soburi commented:
@simonguinot 

> So how does this frequency affect the pulses ?

The frequency defines the duration to transmit 1bit data. The ratio of high/low are not changed.
If the frequency changes to x2, The pulse width became 1/2.

> Let's say a user selects the 400kHz frequency for a strip of WS2811 devices. In that case is it still correct for the PIO program to output 0001100000 and 0001111111 pulses for 0 and 1 bits ?
> 
> Did you consider to make the pulses configurable as in the SPI driver (see the spi-one-frame and spi-zero-frame DT properties) ? This would make this driver compatible with more LED controllers.

Added waveform configuration feature.
Described in ...
https://github.com/zephyrproject-rtos/zephyr/pull/55226/files#diff-fcf73ed565033a3d9aaa31c7e85daa8b76f937e904198109ef74b06412b60a8fR14-R44

soburi commented:
@simonguinot @mbolivar-ampere 
Could you take a look please?

simonguinot reviewed (commented):
Hi @soburi,

Sorry again for the delay. It took me some time to break into this driver.

Please see my main concern below. I'll look into the details once this will be resolved.

soburi commented on dts/bindings/led_strip/raspberrypi,pico-pio-ws2812.yaml:
```diff
@@ -0,0 +1,44 @@
+description: |
+  The pio node configured for ws2812.
+
+compatible: "raspberrypi,pico-pio-ws2812"
```
It comes from rpi_pico's PIO spec.

- The PIO can load a program.
- The PIO can run the program as four independent instances.
- Each instance can set frequency separately.

The LED strip driver corresponds to a single LED strip.
For example, creating four driver nodes in dts can drive 4 LED strips.
The instances that drive the LED strips can configure the frequency independently for each.
But the `waveform` (It means the ratio of L and H here) is tied to the PIO program. So, All instances use the same configuration.

In other word, "In other words, there is a 1:N relationship between PIO and LED strip".
My design is mapping those hardware capabilities simply dts.

soburi replied:
> OK we can have 4 PIO instances running at different frequencies but programmed with the same waveform. So the driver can only manage up to 4 LED strips IF they are of the same model. Different models require different waveforms. Am I good here ?

Thank you for reviewing.
The comment you wrote is right. I understand with your concern.
The current design prioritizes being able to perform all possible hardware settings.
From the perspective of ease of understanding, it is better to set all the frequencies to be the same, as you said.

soburi replied:
This is left as previously.

soburi commented on drivers/led_strip/ws2812_rpi_pico_pio.c:
```diff
@@ -0,0 +1,250 @@
+	const struct pio_program program;
+};
+
+#ifdef CONFIG_DT_HAS_WORLDSEMI_WS2812_RPI_PICO_PIO_ENABLED
```
> Well, is there a point of having this whole driver enabled at all if this DT property is missing ? Without, the driver only adds a PIO program but nothing more...

Yes, it does nothing. But I think it is good to compile the PIO part if enabled in dts from the consistency viewpoint.
In such cases, This part causes a compiler warning.

> This comment leads back to my previous question: why are we now handling two compatibles in this driver ?


It is what I wrote in the previous comment. Because PIO and LED strip relation are not 1:1. It is better not to combine into one. But it is used only at the same time. So I think it may be better not to split file.

soburi replied:
2023/11/19

Splitting PIO part into rpi_pico_pio_ws2812.c.

soburi commented:
@mbolivar-nordic @mbolivar-ampere 
Could you take a look, please?

thedjnK reviewed (dismissed):
Works perfect on rpi pico!

simonguinot commented:
> @mbolivar-nordic @mbolivar-ampere Could you take a look, please?

Hi @soburi. I apologize for the delay to get this PR merged. It is not OK.

I suggest we move on without @mbolivar-ampere. He is busy and he'll catch up eventually.

So please address my comments and let's get this merged.

soburi commented:
@simonguinot 

> Hi @soburi. I apologize for the delay to get this PR merged. It is not OK.
> I suggest we move on without @mbolivar-ampere. He is busy and he'll catch up eventually.
> So please address my comments and let's get this merged.

I addressed or responded to the comments I can see.
It seems that some comments are no longer visible due to the update.
I do my best to accommodate you, but please check.

simonguinot commented:
> @mbolivar-nordic @mbolivar-ampere Could you take a look, please?

Hi @soburi. I apologize for the delay to get this PR merged. It is not OK.

I suggest we move on without @mbolivar-ampere. He is busy and he'll catch up eventually.

So please address my comments and let's get this merged.

> @simonguinot
> 
> > Hi @soburi. I apologize for the delay to get this PR merged. It is not OK.
> > I suggest we move on without @mbolivar-ampere. He is busy and he'll catch up eventually.
> > So please address my comments and let's get this merged.
> 
> I addressed or responded to the comments I can see. It seems that some comments are no longer visible due to the update. I do my best to accommodate you, but please check.

Great ! I'll look at it this week.

simonguinot reviewed (changes requested):
Hi @soburi,

I am not sure with the design of the DT bindings. They seems confusing at best. Since I am not an expert in DT (from far), let's try to find some help to review them.

simonguinot commented:
@soburi I suggest you implement the following DT design:

```
ws2812 {
        compatible = "worldsemi,ws2812-rpi_pico-pio";
        status = "okay";

        ...

        bit-waveform = ... ;

        ws2812_1 {
                status = "okay"; 

                output-pin = ... ; 
                frequency = ...;

                chain-length = ... ;
                color-mapping = ... ;
                reset-delay = ... ;
        };
        ws2812_2 {
                status = "okay";

                output-pin = ... ;
                frequency = ...;

                chain-length = ... ;
                color-mapping = ... ;
                reset-delay = ... ;
        };
        ...
};
```

And in the driver you can create a device for each child node by using `DT_INST_FOREACH_CHILD_STATUS_OKAY`.

I think it would be less confusing than the design you are proposing and I would be OK with that.

simonguinot commented:
> @soburi I suggest you implement the following DT design:
> 
> ```
> ws2812 {
>         compatible = "worldsemi,ws2812-rpi_pico-pio";
>         status = "okay";
> 
>         ...
> 
>         bit-waveform = ... ;
> 
>         ws2812_1 {
>                 status = "okay"; 
> 
>                 output-pin = ... ; 
>                 frequency = ...;
> 
>                 chain-length = ... ;
>                 color-mapping = ... ;
>                 reset-delay = ... ;
>         };
>         ws2812_2 {
>                 status = "okay";
> 
>                 output-pin = ... ;
>                 frequency = ...;
> 
>                 chain-length = ... ;
>                 color-mapping = ... ;
>                 reset-delay = ... ;
>         };
>         ...
> };
> ```
> 
> And in the driver you can create a device for each child node by using `DT_INST_FOREACH_CHILD_STATUS_OKAY`.
> 
> I think it would be less confusing than the design you are proposing and I would be OK with that.

What do you think @soburi ?

simonguinot commented:
Hi @rexut,
Please can you do something about the commits referencing this PR (from another GitHub repository) ? This makes really difficult to follow reviewers' comments. Thanks in advance.

rexut commented:
> Hi @rexut, Please can you do something about the commits referencing this PR (from another GitHub repository) ? This makes really difficult to follow reviewers' comments. Thanks in advance.

Oh god, I'm so sorry. I'll remove the reference ASAP. Thx, for that hint. This is really very bad behavior on Github's part.

soburi commented:
> > And in the driver you can create a device for each child node by using `DT_INST_FOREACH_CHILD_STATUS_OKAY`.
> > I think it would be less confusing than the design you are proposing and I would be OK with that.
> 
> What do you think @soburi ?

Great suggestion!
Now I can reorganize the PIO and ws2812 drivers neatly.
I've rewritten this PR accordingly for this suggestion.

soburi commented on drivers/led_strip/rpi_pico_pio_ws2812.c:
```diff
@@ -0,0 +1,87 @@
+
+#define LOG_LEVEL CONFIG_LED_STRIP_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(rpi_pico_pio_ws2812);
```
Thank you for reviewing.
I applied this suggestion.

soburi commented on drivers/led_strip/Kconfig.ws2812:
```diff
@@ -46,4 +46,12 @@ config WS2812_STRIP_GPIO
+	depends on DT_HAS_WORLDSEMI_WS2812_RPI_PICO_PIO_ENABLED
+	select PICOSDK_USE_PIO
+	help
+	  The WS2812 LED strip driver that using RaspberryPi Pico's PIO feature.
```
I applied this suggestion.

soburi commented on drivers/led_strip/rpi_pico_pio_ws2812.c:
```diff
@@ -0,0 +1,87 @@
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/drivers/misc/pio_rpi_pico/pio_rpi_pico.h>
+#include <zephyr/sys_clock.h>
+#include <zephyr/kernel.h>
```
For using `k_usleep`, it is required to include `zephyr/kernel.h`.
However, `zephyr/device.h` and `zephyr/sys_clock.h` are not needed here.
I removed it.

soburi commented on drivers/led_strip/ws2812_rpi_pico_pio.c:
```diff
@@ -0,0 +1,179 @@
+
+#define LOG_LEVEL CONFIG_LED_STRIP_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(ws2812_rpi_pico_pio);
```
Fixed.

soburi commented on drivers/led_strip/CMakeLists.txt:
```diff
@@ -7,4 +7,5 @@ zephyr_library_sources_ifdef(CONFIG_LPD880X_STRIP       lpd880x.c)
 zephyr_library_sources_ifdef(CONFIG_WS2812_STRIP_GPIO   ws2812_gpio.c)
 zephyr_library_sources_ifdef(CONFIG_WS2812_STRIP_SPI    ws2812_spi.c)
 zephyr_library_sources_ifdef(CONFIG_WS2812_STRIP_I2S    ws2812_i2s.c)
+zephyr_library_sources_ifdef(CONFIG_WS2812_STRIP_RPI_PICO_PIO	ws2812_rpi_pico_pio.c rpi_pico_pio_ws2812.c)
```
When I reorganized it, I made only ws2812_rpi_pico_pio.c.

soburi commented on dts/bindings/led_strip/worldsemi,ws2812-rpi_pico-pio.yaml:
```diff
@@ -0,0 +1,28 @@
+# Copyright (c) 2023, TOKITA Hiroshi
```
Oops! Fixed it.

soburi commented:
@simonguinot 

Sorry for the very slow response.

I reorganized it based on your proposal at https://github.com/zephyrproject-rtos/zephyr/pull/55226#issuecomment-1848408903.
I think this has organized the files and made them easier to understand.
thank you.


I'm also responding to your comments.
I apologize for the inconvenience, but please confirm.

https://github.com/zephyrproject-rtos/zephyr/pull/55226#issuecomment-1848408903](https://github.com/zephyrproject-rtos/zephyr/pull/55226#issuecomment-1821916981)

---
P.S.

I was a little busy because I had an announcement at Open Source Summit Japan.
https://ossjapan2023.sched.com/event/1Tyqu/supporting-hobbyist-friendly-opensource-hardware-in-zephyr-hiroshi-tokita-fujitsu-ltd?iframe=no&w=100%25&sidebar=yes&bg=no&fbclid=IwAR2uLP_mH_qeQajl9LlBiIfSbR7BwHdEioNLtkEt6 QxZiVBsUmXlHhfD2xc

I also gave a little introduction to this patch.
Thank you for your long-term review.

rexut commented on dts/bindings/led_strip/worldsemi,ws2812-rpi_pico-pio.yaml:
**suggestion (non-blocking):** The new compatibility string is somewhat misleading and could suggest that this feature could only be used for one board. The name "rpi_pico" refers to a board, namely the "Raspberry Pi Pico", which is based on the SoC called "RP2040". It is precisely this SoC that has this special PIO unit, not the board. A naming "worldsemi,ws2812-rp2040-pio" would be more accurate and comprehensible here.

soburi replied:
This has been a topic of discussion since the beginning of support for rpi pico, but at this point the official name of the SoC family name is unknown, and the discussion will be reconsidered once a successor model is released and the family name is confirmed. This PR also follows that policy.


https://discord.com/channels/720317445772017664/938474761405726800/1177014983423442964

> This is the discussion at the time of the first commit:
> https://github.com/zephyrproject-rtos/zephyr/pull/34835#pullrequestreview-735878199
> The conclusion is that the name of a clear "SoC family" is unknown now.
> There isn't clear consistency, including Pico-SDK.
> 
> The introduction of RP1 has made things even more confusing.
> https://datasheets.raspberrypi.com/rp1/rp1-peripherals.pdf
> 
> I think it would be okay to reconsider when the family name becomes clearer as variations increase, as was the case in the original discussion.

rexut replied:
Okay, I understand. I didn't realize how confusing the current situation with the chip name is. I had been wondering about it for months…

rexut reviewed (changes requested):
@soburi first of all, thank you very much for all your effort to work on this feature. You were my savior when I found a WS2812 LED on a "non-SPI" pin and only the PIO of the RP2040 could be used.

Please have a look at my little comment. If a small renaming is still possible, it would be good.

simonguinot commented:
> @simonguinot
> 
> Sorry for the very slow response.
> 
> I reorganized it based on your proposal at [#55226 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/55226#issuecomment-1848408903). I think this has organized the files and made them easier to understand. thank you.
> 
> I'm also responding to your comments. I apologize for the inconvenience, but please confirm.
> ## [#55226 (review)](https://github.com/zephyrproject-rtos/zephyr/pull/55226#pullrequestreview-1748149446)
> 
> P.S.
> 
> I was a little busy because I had an announcement at Open Source Summit Japan. https://ossjapan2023.sched.com/event/1Tyqu/supporting-hobbyist-friendly-opensource-hardware-in-zephyr-hiroshi-tokita-fujitsu-ltd?iframe=no&w=100%25&sidebar=yes&bg=no&fbclid=IwAR2uLP_mH_qeQajl9LlBiIfSbR7BwHdEioNLtkEt6 QxZiVBsUmXlHhfD2xc
> 
> I also gave a little introduction to this patch. Thank you for your long-term review.

Hi @soburi,



> @simonguinot
> 
> Sorry for the very slow response.
> 
> I reorganized it based on your proposal at [#55226 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/55226#issuecomment-1848408903). I think this has organized the files and made them easier to understand. thank you.
> 
> I'm also responding to your comments. I apologize for the inconvenience, but please confirm.
> ## [#55226 (review)](https://github.com/zephyrproject-rtos/zephyr/pull/55226#pullrequestreview-1748149446)
> 
> P.S.
> 
> I was a little busy because I had an announcement at Open Source Summit Japan. https://ossjapan2023.sched.com/event/1Tyqu/supporting-hobbyist-friendly-opensource-hardware-in-zephyr-hiroshi-tokita-fujitsu-ltd?iframe=no&w=100%25&sidebar=yes&bg=no&fbclid=IwAR2uLP_mH_qeQajl9LlBiIfSbR7BwHdEioNLtkEt6 QxZiVBsUmXlHhfD2xc
> 
> I also gave a little introduction to this patch. Thank you for your long-term review.

Hi @soburi,

Thanks for the update. I'll review it this week-end.

P.S. Nice talk at the Open Source Summit Japan !

simonguinot commented on dts/bindings/led_strip/worldsemi,ws2812-rpi_pico-pio.yaml:
```diff
@@ -0,0 +1,69 @@
+    description: |
+      This property defines the waveform for sending 1-bit data.
+      The program uses the first three elements of the array.
+      Where, let these value be T0, T1, and T2.
```
values ^
And I am not sure about the "Where, " part. You should probably remove it.

soburi replied:
Removed the "Where, " part.

simonguinot reviewed (commented):
Thanks for your patience @soburi. Great work ! It looks good to me. See below for my only comment.

Let's try to bring more eyes on this PR in order to get one more approval.

simonguinot commented:
Hi @bbilas. Please can you have a look at this PR ? Thanks in advance !

bbilas commented on drivers/led_strip/Kconfig.ws2812:
```diff
@@ -46,4 +46,11 @@ config WS2812_STRIP_GPIO
 
+config WS2812_STRIP_RPI_PICO_PIO
+	bool "Raspberry Pi Pico PIO"
+	depends on DT_HAS_WORLDSEMI_WS2812_RPI_PICO_PIO_ENABLED
```
You can add

<pre>
default y
</pre>

since it depends on the dts node status value.

soburi replied:
Changed it to DBG.

simonguinot replied:
As a side note the `led_strip` drivers have not been converted to the DT node availability + `default y` mechanism.

In addition we had some complains about the automagic LED driver selection: https://github.com/zephyrproject-rtos/zephyr/issues/54609.

bbilas commented on drivers/led_strip/ws2812_rpi_pico_pio.c:
```diff
@@ -0,0 +1,248 @@
+static int ws2812_led_strip_update_channels(const struct device *dev, uint8_t *channels,
+					    size_t num_channels)
+{
+	LOG_ERR("update_channels not implemented");
```
I would decrease the log level to DBG or even remove it completely.

soburi replied:
Changed it to DBG.

bbilas commented on drivers/led_strip/ws2812_rpi_pico_pio.c:
```diff
@@ -0,0 +1,248 @@
+	const struct ws2812_led_strip_config *config = dev->config;
+	struct ws2812_led_strip_data *data = dev->data;
+	int sm;
+
```
Would not be better to move the following part

<pre>
if (!device_is_ready(config->piodev)) {
		LOG_ERR("%s: PIO device not ready", dev->name);
		return -ENODEV;
	}
</pre>

right here from `ws2812_led_strip_sm_init` function? We would be able to catch not initialized device pointer before that for loop.

soburi replied:
Applied this suggestion.

bbilas commented on drivers/led_strip/ws2812_rpi_pico_pio.c:
```diff
@@ -0,0 +1,248 @@
+	if (sm < 0) {
+		return sm;
+	}
+	data->sm = sm;
```
Please add nl before that.

soburi replied:
Inserted empty line.

bbilas commented on samples/drivers/led_ws2812/boards/adafruit_kb2040.conf:
```diff
@@ -0,0 +1 @@
+CONFIG_WS2812_STRIP_RPI_PICO_PIO=y
```
If you enable the driver by default, this will not be necessary.

soburi replied:
WS2812_STRIP_DRIVER is the `choice`, so this must be left.

bbilas replied:
My bad, in that case, I would move it into the board's defconfig. @simonguinot what do you think about it?

simonguinot reviewed (approved):
Thanks for the great driver @soburi !

simonguinot commented:
Thanks @bbilas !

soburi commented:
> Thanks for the great driver @soburi !

Deeply thanks for your long-term review. I'm relieved that the PR was approved.

kartben commented:
Can we now get a Zephyr-based alternative to [WLED](https://kno.wled.ge/), please? 😬

soburi commented:
@kartben 

> Can we now get a Zephyr-based alternative to [WLED](https://kno.wled.ge/), please? 😬

I think It is possible with wired ethernet. such as https://docs.zephyrproject.org/latest/boards/arm/w5500_evb_pico/doc/index.html.

Still, there is a major obstacle to implementing this with wireless LAN.
https://github.com/zephyrproject-rtos/zephyr/issues/67092

simonguinot commented:
> Can we now get a Zephyr-based alternative to [WLED](https://kno.wled.ge/), please? 😬

Indeed, on the top of the LED (and LED strip) driver API, we are missing a Zepyr userland module to provides LED effects and expose an API for remote control through network, USB, I2C  (or other busses). And also we would like to make it compatible enough with existing APIs such as Razer or Microsoft Dynamic Lighting.

thedjnK commented:
> > Can we now get a Zephyr-based alternative to [WLED](https://kno.wled.ge/), please? 😬
> 
> Indeed, on the top of the LED (and LED strip) driver API, we are missing a Zepyr userland module to provides LED effects and expose an API for remote control through network, USB, I2C (or other busses). And also we would like to make it compatible enough with existing APIs such as Razer or Microsoft Dynamic Lighting.

There is this which is very primitive: https://github.com/thedjnK/adalight_rgb but allows usage of https://github.com/psieg/Lightpack for doing effects on LED strips from a PC
